controller/controller.go

```diff
@@ -220,14 +220,33 @@ func (c *Controller) RunOnce(ctx context.Context) error {
 		ManagedRecords:     c.ManagedRecordTypes,
 	}
 
-	plan = plan.Calculate()
+	p = p.Calculate()
 
-	if plan.Changes.HasChanges() {
-		err = c.Registry.ApplyChanges(ctx, plan.Changes)
+	if p.Changes.HasChanges() {
+		err = c.Registry.ApplyChanges(ctx, p.Changes)
 		if err != nil {
-			registryErrorsTotal.Inc()
-			deprecatedRegistryErrors.Inc()
-			return err
+			// if the original batch fails, try each action
+			createOnlyPolicy := &plan.CreateOnlyPolicy{}
+			createChanges := createOnlyPolicy.Apply(p.Changes)
+			createErr := c.Registry.ApplyChanges(ctx, createChanges)
+			// if the single action batch fails, try batch in halves
+			if createErr != nil {
+				applyHalfActionChanges(c, ctx, createChanges)
+			}
+
+			updateOnlyPolicy := &plan.UpdateOnlyPolicy{}
+			updateChanges := updateOnlyPolicy.Apply(p.Changes)
+			updateErr := c.Registry.ApplyChanges(ctx, updateChanges)
+			if updateErr != nil {
+				applyHalfActionChanges(c, ctx, updateChanges)
+			}
+
+			deleteOnlyPolicy := &plan.DeleteOnlyPolicy{}
+			deleteChanges := deleteOnlyPolicy.Apply(p.Changes)
+			deleteErr := c.Registry.ApplyChanges(ctx, deleteChanges)
+			if deleteErr != nil {
+				applyHalfActionChanges(c, ctx, deleteChanges)
+			}
```

Comment:
Looks good 👍 

Maybe we can also add a textual description of the approach taken as a comment or in the PR. Maybe it helps us and other reviewers to reason about whether this splitting is safe and doesn't change the final result (apart from ignoring the failing records).

Reply:
yes, yes. the current comments are more like self notes to assimilate what I'm doing in the draft, once it's ready for opening a PR upstream there will more clear documentation.